pandas/core/indexes/base.py

```diff
@@ -6262,7 +6262,11 @@ def _should_compare(self, other: Index) -> bool:
             return False
 
         dtype = _unpack_nested_dtype(other)
-        return self._is_comparable_dtype(dtype) or is_object_dtype(dtype)
+        return (
+            self._is_comparable_dtype(dtype)
+            or is_object_dtype(dtype)
+            or is_string_dtype(dtype)
```

Comment:
why is the is_string_dtype case needed here?

Reply:
Because we allow to compare non-strings with strings for certain cases, I guess? 

```
In [4]: pd.date_range("2020-01-01", periods=3) == "2020-01-01"
Out[4]: array([ True, False, False])
```

Reply:
hmm looks like two tests fail if this is disabled, test_map_missing_mixed, test_get_indexer_datetime.  The map test I'd have to look into, but the get_indexer one make sense:

```
ii = IntervalIndex.from_breaks(date_range("2018-01-01", periods=4))
target = DatetimeIndex(["2018-01-02"], dtype="M8[ns]")
result = ii.get_indexer(target.astype(str))
> tm.assert_numpy_array_equal(result, expected)
```

I suspect there are dtype combinations for which we don't need the is_string_dtype check here, might be able to move it to _is_comparable_dtype and restore fast-paths in some cases.